src/hb-ot-layout.cc

```diff
@@ -1989,6 +2058,184 @@ hb_ot_layout_get_baseline (hb_font_t                   *font,
+ * @baseline_tag: a baseline tag
+ * @direction: text direction.
+ * @script_tag:  script tag.
+ * @language_tag: language tag, currently unused.
+ * @coord: (out): baseline value if found.
+ *
+ * Fetches a baseline value from the face, and synthesizes
+ * it if the font does not have it.
+ *
+ * Since: REPLACEME
+ **/
+void
+hb_ot_layout_get_baseline_with_fallback (hb_font_t                   *font,
+					 hb_ot_layout_baseline_tag_t  baseline_tag,
+					 hb_direction_t               direction,
+					 hb_tag_t                     script_tag,
+					 hb_tag_t                     language_tag,
+					 hb_position_t               *coord /* OUT */)
+{
+  if (hb_ot_layout_get_baseline (font,
+                                 baseline_tag,
+                                 direction,
+                                 script_tag,
+                                 language_tag,
+                                 coord))
+    return;
+
+  /* Synthesize missing baselines.
+   * See https://www.w3.org/TR/css-inline-3/#baseline-synthesis-fonts
+   */
+  switch ((int) baseline_tag)
+  {
+  case HB_OT_LAYOUT_BASELINE_TAG_ROMAN:
+    *coord = 0;
+    break;
+
+  case HB_OT_LAYOUT_BASELINE_TAG_MATH:
+    {
+      hb_codepoint_t glyph;
+      hb_glyph_extents_t extents;
+      if (HB_DIRECTION_IS_HORIZONTAL (direction) &&
+          (hb_font_get_nominal_glyph (font, 0x2212u, &glyph) ||
+           hb_font_get_nominal_glyph (font, '-', &glyph)) &&
+          hb_font_get_glyph_extents (font, glyph, &extents))
+      {
+        *coord = extents.y_bearing + extents.height / 2;
+      }
+      else
+      {
+        hb_position_t x_height = 0;
+        hb_ot_metrics_get_position (font, HB_OT_METRICS_TAG_X_HEIGHT, &x_height);
+        *coord = x_height / 2;
+      }
+    }
+    break;
+
+  case HB_OT_LAYOUT_BASELINE_TAG_IDEO_FACE_TOP_OR_RIGHT:
+  case HB_OT_LAYOUT_BASELINE_TAG_IDEO_FACE_BOTTOM_OR_LEFT:
+    {
+      hb_position_t embox_top, embox_bottom;
+
+      hb_ot_layout_get_baseline_with_fallback (font,
+                                               HB_OT_LAYOUT_BASELINE_TAG_IDEO_EMBOX_TOP_OR_RIGHT,
+                                               direction,
+                                               script_tag,
+                                               language_tag,
+                                               &embox_top);
+      hb_ot_layout_get_baseline_with_fallback (font,
+                                               HB_OT_LAYOUT_BASELINE_TAG_IDEO_EMBOX_BOTTOM_OR_LEFT,
+                                               direction,
+                                               script_tag,
+                                               language_tag,
+                                               &embox_bottom);
+
+      if (baseline_tag == HB_OT_LAYOUT_BASELINE_TAG_IDEO_FACE_TOP_OR_RIGHT)
+        *coord = embox_top + (embox_bottom - embox_top) * 2 / 10;
+      else
+        *coord = embox_bottom + (embox_top - embox_bottom) * 2 / 10;
```

Comment:
Why this particular constant of 20%?

Reply:
I think we agreed here that 20% is too much. Its now 10%